lightning/src/ln/peers/handshake/mod.rs

```diff
@@ -2,85 +2,34 @@
 //! Handshake states can be advanced automatically, or by manually calling the appropriate step.
 //! Once complete, returns an instance of Conduit.
 
-use bitcoin::secp256k1;
-
-use bitcoin::hashes::{Hash, HashEngine};
-use bitcoin::hashes::sha256::Hash as Sha256;
 use bitcoin::secp256k1::{PublicKey, SecretKey};
 
-use ln::peers::{chacha, hkdf};
-use ln::peers::conduit::{Conduit, SymmetricKey};
-use ln::peers::handshake::acts::{ActOne, ActThree, ActTwo, ACT_ONE_LENGTH, ACT_TWO_LENGTH, ACT_THREE_LENGTH, Act};
-use ln::peers::handshake::hash::HandshakeHash;
-use ln::peers::handshake::states::{ActOneExpectation, ActThreeExpectation, ActTwoExpectation, HandshakeState};
+use ln::peers::conduit::Conduit;
+use ln::peers::handshake::states::{HandshakeState, IHandshakeState};
 
-pub(crate) mod acts;
-mod hash;
 mod states;
-mod tests;
 
 /// Object for managing handshakes.
 /// Currently requires explicit ephemeral private key specification.
 pub struct PeerHandshake {
 	state: Option<HandshakeState>,
-	private_key: SecretKey,
-	remote_public_key: Option<PublicKey>,
-	ephemeral_private_key: SecretKey,
-
-	read_buffer: Vec<u8>,
 }
 
 impl PeerHandshake {
 	/// Instantiate a new handshake with a node identity secret key and an ephemeral private key
-	pub fn new_outbound(private_key: &SecretKey, remote_public_key: &PublicKey, ephemeral_private_key: &SecretKey) -> Self {
-		Self {
-			state: Some(HandshakeState::Uninitiated),
-			private_key: (*private_key).clone(),
-			remote_public_key: Some(remote_public_key.clone()),
-			ephemeral_private_key: (*ephemeral_private_key).clone(),
-			read_buffer: Vec::new(),
-		}
-	}
-
-	/// Instantiate a new handshake in anticipation of a peer's first handshake act
-	pub fn new_inbound(private_key: &SecretKey, ephemeral_private_key: &SecretKey) -> Self {
-		let mut handshake = Self {
-			state: Some(HandshakeState::Uninitiated),
-			private_key: (*private_key).clone(),
-			remote_public_key: None,
-			ephemeral_private_key: (*ephemeral_private_key).clone(),
-			read_buffer: Vec::new(),
-		};
-		let public_key = Self::private_key_to_public_key(&private_key);
-		let (hash, chaining_key) = Self::initialize_state(&public_key);
-		handshake.state = Some(HandshakeState::AwaitingActOne(ActOneExpectation {
-			hash,
-			chaining_key,
-		}));
-		handshake
-	}
+	pub fn create_and_initialize_outbound(initiator_static_private_key: &SecretKey, responder_static_public_key: &PublicKey, initiator_ephemeral_private_key: &SecretKey) -> (Vec<u8>, Self) {
```

Comment:
The main alternative I thought about was the `new()/setup()` pattern. In that approach, `peer_handler` just needs to know that it should call `setup()` after construction and not expose the empty vector `process_act()`. It may seem unnecessary to add a pass-through function in `PeerHandshake` that just calls `process_act(&[])`, but it makes the layering a bit saner and hides a bit more of the implementation details.

```
handshake = PeerHandshake::new_outbound(...);
let bytes = handshake.setup();
```

The` PeerHandshake` is so minimal that I wouldn't even advocate for additional unit tests for that case since the non-empty vector case against `InitiatorStartingState` should highlight an issue if the 1 use is ever changed.

I haven't found a great way to do this type of pattern and am flexible on the approach. I'm sure something like it will come up again so having a standard pattern to point to would be great.

Reply:
I think I actually really like the setup approach, but this is personal preference, and I think we should ultimately defer to @ariard and @thebluematt.

Reply:
> I haven't found a great way to do this type of pattern and am flexible on the approach. I'm sure something like it will come up again so having a standard pattern to point to would be great.

The example I've in mind is `Channel::new_outbound/get_open_channel` for channel establishement, first creating the object channel and then allowing to extract a `OpenChannel` to send to will-be counterparty. So it binds towards the splitting done in https://github.com/arik-so/rust-lightning/pull/2/commits/5d951fb9b2213d33bc6312a324ef6b5a345d4daf. 

In that present case, I don't have a strong opinion, beyond a general principle of favoring flexibility.

Reply:
Resolved: I split this out into a set_up_outbound() function.